### PR TITLE
compile-time evaluation: detect writes through immutable pointers

### PR DESCRIPTION
**Pull request by RalfJung**
This has two motivations:
- it unblocks https://github.com/rust-lang/rust/pull/116745 (and therefore takes a big step towards `const_mut_refs` stabilization), because we can now detect if the memory that we find in `const` can be interned as "immutable"
- it would detect the UB that was uncovered in https://github.com/rust-lang/rust/pull/117905, which was caused by accidental stabilization of `copy` functions in `const` that can only be called with UB

When UB is detected, we emit a future-compat warn-by-default lint. This is not a breaking change, so completely in line with [the const-UB RFC](https://rust-lang.github.io/rfcs/3016-const-ub.html), meaning we don't need t-lang FCP here. I made the lint immediately show up for dependencies since it is nearly impossible to even trigger this lint without `const_mut_refs` -- the accidentally stabilized `copy` functions are the only way this can happen, so the crates that popped up in #117905 are the only causes of such UB (in the code that crater covers), and the three cases of UB that we know about have all been fixed in their respective crates already.

The way this is implemented is by making use of the fact that our interpreter is already generic over the notion of provenance. For CTFE we now use the new `CtfeProvenance` type which is conceptually an `AllocId` plus a boolean `immutable` flag (but packed for a more efficient representation). This means we can mark a pointer as immutable when it is created as a shared reference. The flag will be propagated to all pointers derived from this one. We can then check the immutable flag on each write to reject writes through immutable pointers.

I just hope perf works out.

### CODE REVIEW

File: compiler/rustc_middle/src/mir/interpret/pointer.rs [hunk 1]
```diff
@@ -141,6 +141,78 @@ pub trait Provenance: Copy + fmt::Debug + 'static {
...
+    }
+}
+
+const IMMUTABLE_MASK: u64 = 1 << 63; // the highest bit
+
+impl CtfeProvenance {
+    /// Returns the `AllocId` of this provenance.
+    #[inline(always)]
+    pub fn alloc_id(self) -> AllocId {
+        AllocId(NonZeroU64::new(self.0.get() & !IMMUTABLE_MASK).unwrap())
```
Comment (RalfJung) [inline compiler/rustc_middle/src/mir/interpret/pointer.rs:170]: We could use` new_unchecked` here, we have an invariant that the AllocId part of the provenance is always non-null. But let's first see if perf is fine without unsafe code.

File: compiler/rustc_middle/src/mir/interpret/allocation.rs [hunk 2]
```diff
@@ -336,14 +336,14 @@ impl<Prov: Provenance, Bytes: AllocBytes> Allocation<Prov, (), Bytes> {
     }
 }
 
-impl<Bytes: AllocBytes> Allocation<AllocId, (), Bytes> {
+impl<Bytes: AllocBytes> Allocation<CtfeProvenance, (), Bytes> {
     /// Adjust allocation from the ones in tcx to a custom Machine instance
     /// with a different Provenance and Extra type.
```
Comment (metreno) [inline compiler/rustc_middle/src/mir/interpret/allocation.rs:341]:
```suggestion
    /// Adjust allocation from the ones in `tcx` to a custom Machine instance
    /// with a different `Provenance `and `Extra `type.
```

File: compiler/rustc_middle/src/mir/interpret/allocation/provenance_map.rs [hunk 3]
```diff
@@ -6,13 +6,13 @@ use std::cmp;
...
-use super::{alloc_range, AllocError, AllocId, AllocRange, AllocResult, Provenance};
+use super::{alloc_range, AllocError, AllocRange, AllocResult, CtfeProvenance, Provenance};
 use rustc_serialize::{Decodable, Decoder, Encodable, Encoder};
 
 /// Stores the provenance information of pointers stored in memory.
 #[derive(Clone, PartialEq, Eq, Hash, Debug)]
 #[derive(HashStable)]
-pub struct ProvenanceMap<Prov = AllocId> {
+pub struct ProvenanceMap<Prov = CtfeProvenance> {
     /// Provenance in this map applies from the given offset for an entire pointer-size worth of
```
Comment (metreno) [inline compiler/rustc_middle/src/mir/interpret/allocation/provenance_map.rs:16]:
```suggestion
    /// `Provenance `in this map applies from the given offset for an entire pointer-size worth of
```

File: compiler/rustc_lint_defs/src/builtin.rs [hunk 4]
```diff
@@ -8,6 +8,135 @@ use crate::{declare_lint, declare_lint_pass, FutureIncompatibilityReason};
 use rustc_span::edition::Edition;
 use rustc_span::symbol::sym;
 
+declare_lint_pass! {
```
Comment (RalfJung) [inline compiler/rustc_lint_defs/src/builtin.rs:11]: I moved this to the top, it was in a random spot in the middle of the file which surely doesn't make a ton of sense.

**Comment by RalfJung** [comment]
Right, so there is a regression in ctfe-stress. I was worried this might happen. I guess now comes the usual phase of doing random changes and seeing how they affect perf...

**Comment by RalfJung** [comment]
Oh, somehow that's worse...

I guess I'll need to get a proper callgrind diff. I only wish debian would finally update its valgrind so this wouldn't be so painful... :( not sure when I'll have the time to deal with this.

**Comment by RalfJung** [comment]
Yeah if you could do that for one of the "full" builds in https://perf.rust-lang.org/compare.html?start=0b8a61b235662d397721d1b88ddefdfc147ba39a&end=b5c17f82cf090a61e41e7d3c33a8975322d8d5c3&stat=instructions:u that'd be great. My valgrind can't handle the DWARF 5 stuff...

**Comment by saethlin** [comment]
[cgann-diff-0b8a61b235662d397721d1b88ddefdfc147ba39a-b5c17f82cf090a61e41e7d3c33a8975322d8d5c3-ctfe-stress-5-Opt-Full.txt](https://github.com/rust-lang/rust/files/13477160/cgann-diff-0b8a61b235662d397721d1b88ddefdfc147ba39a-b5c17f82cf090a61e41e7d3c33a8975322d8d5c3-ctfe-stress-5-Opt-Full.txt)

**Comment by RalfJung** [comment]
So that's the entire diff?
```
866,604,465  ???:rustc_const_eval::const_eval::eval_queries::eval_body_using_ecx
 53,052,961  ???:<rustc_middle::ty::Ty>::is_trivially_freeze
 -4,718,594  ???:<rustc_middle::ty::context::TyCtxt>::const_eval_resolve
```
Hm, that's not terribly helpful...

**Comment by RalfJung** [comment]
I guess the `is_trivially_freeze` showing up is a hint though -- we are now determining whether the pointee is `Freeze` each time a shared reference is created. That's probably the source of the regression. I don't think it is avoidable. There already is a `is_trivially_freeze` fast-path so I don't see ways to optimize that further.

**Comment by RalfJung** [comment]
Yeah, none of these smaller changes matter. That's all just `is_freeze` overhead. It (almost) only shows up in the ctfe stress test so... I guess it's fine. Not great but it's worth it for the UB detection and interner improvements IMO.

**Comment by RalfJung** [comment]
Hm, actually... one last idea. Let's skip `is_freeze` when the pointer is already immutable.

@bors try @rust-timer queue

**Comment by saethlin** [comment]
Nice!

**Comment by RalfJung** [comment]
Indeed. :) Though unfortunately that didn't make the regression any smaller for the non-ctfe-stress tests.

**Comment by saethlin** [comment]
:shrug: I feel like the remaining regression is small enough to be acceptable. We're doing more checks in const-eval, and there are no regressions in the builds of real crates.

File: compiler/rustc_middle/src/mir/interpret/pointer.rs [hunk 5]
```diff
@@ -141,6 +126,80 @@ pub trait Provenance: Copy + fmt::Debug + 'static {
...
 
+/// The type of provenance in the compiler-time interpreter.
+/// This is a packed representation of an `AllocId` and an `immutable: bool`.
+#[derive(Copy, Clone, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub struct CtfeProvenance(NonZeroU64);
+
+impl From<AllocId> for CtfeProvenance {
+    fn from(value: AllocId) -> Self {
+        let prov = CtfeProvenance(value.0);
+        assert!(!prov.immutable(), "`AllocId` with the highest bit set cannot be used in CTFE");
```
Comment (saethlin) [inline compiler/rustc_middle/src/mir/interpret/pointer.rs:137]: Everything else in this PR is mostly just mechanical changes or pretty simple, but I do not understand what we're doing here. Why can we only create a `CtfeProvenance` that we can write through?

Comment (RalfJung) [reply compiler/rustc_middle/src/mir/interpret/pointer.rs:137]: This is asserting that `value` is less than 2^63, to ensure we can use the highest bit to store "immutable".

Comment (saethlin) [reply compiler/rustc_middle/src/mir/interpret/pointer.rs:137]: I see. I would have written this check without going through the accessor, but this is fine. The assert message should have clued me in.

**Comment by saethlin** [comment]
r=me
Leaving you the choice to squash or whatnot.

**Comment by RalfJung** [comment]
I squashed the fixups but kept the commits that are logical units. Thanks for the review!

@bors r=saethlin

**Pull request merged**